### DISCUSSION TITLE
Tutorial 3 challenge - big difference between using dedicated `position` variable and reusing `clip_position`

### DISCUSSION THREAD

**Discussion by Pieresqi** [sotrh/learn-wgpu#498] [Q&A] [answered]
Hello there, 
I am currently going through your `Learn Wgpu` and learning wgpu.

Doing 3rd challenge and comparing it with your solution I have noticed a big difference between using dedicated `position` variable and reusing `clip_position`. Reusing `clip_position` will  give me 1 colored, light yellow, triangle but using dedicated variable will yield multi-colored triangle.

I have changed this https://github.com/sotrh/learn-wgpu/blob/64de216a2d1dccab9cef3739fbc2b076d8616f29/code/beginner/tutorial3-pipeline/src/challenge.wgsl#L24

to this: 
`return vec4<f32>(in.clip_position.xy, 0.5, 1.0);` (or this: `return vec4<f32>(in.clip_position[0], in.clip_position[1], 0.5, 1.0);`, etc)

What's the cause of this big change ?

**Comment by sotrh** [comment, marked as answer]
There is a step in between the vertex shader and the fragment shader that converts the per-vertex data to per-pixel data. The `@builtin(position)` flag corresponds to the final clip position in the vertex shader, but in the fragment shader it is the pixel coordinates of the current fragment.
